### DISCUSSION TITLE
My vcpkg.json is not installing any packages

### DISCUSSION THREAD

**Discussion by frederick-brier** [microsoft/vcpkg#33136] [Q&A] [answered]
I have created a vcpkg.json. However, when running "vcpkg install --dry-run --debug", part of the output is "All requested packages are currently installed." vcpkg list returns "No packages are installed." The vcpkg.json is below. Please tell me what I am doing wrong. Thank you.
```json
{
  "name": "wzlib",
  "version": "1.0.0",
  "dependencies": [
    {
      "name": "boost-thread",
      "version>=": "1.82.0",
      "platform": "(windows & x64 & static)"
    },
    {
      "name": "boost-date-time",
      "version>=": "1.82.0",
      "platform": "(windows & x64 & static)"
    },
    {
      "name": "boost-filesystem",
      "version>=": "1.82.0",
      "platform": "(windows & x64 & static)"
    }
  ],
  "builtin-baseline": "247662ef304453c72acd6b520fa7ff7656e7347c",
  "overrides": [
    {
      "name": "boost-thread",
      "version": "1.67.0"
    },
    {
      "name": "boost-date-time",
      "version": "1.67.0"
    },
    {
      "name": "boost-filesystem",
      "version": "1.67.0"
    }
  ]
}
```

**Comment by Osyotr** [comment]
I don't believe `vcpkg list` works in manifest mode. If `vcpkg install --dry-run` says that everything is already installed then so it is. Check `vcpkg_installed` folder near your manifest file.

**Reply by frederick-brier** [reply]
If a remove vcpkg.json, this command shows all the packages that would be installed:
```
vcpkg install boost-thread boost-filesystem --triplet=x64-windows-static --dry-run --debug --debug-env
```
If I don't remove vcpkg.json, the above command gives an error message saying you cannot combine vcpkg.json with specifying packages on the command line. So it does see the vcpkg.json, it is just that the content is wrong. If there is an error in the vcpkg.json, like an extra comma, it gives an error. It's confusing. I have read everything I can find on vcpkg.json, but something is missing. I have also tried to set all sorts of cmake variables and environment variables:
```cmake
set (VCPKG_MANIFEST_INSTALL ON)
set (VCPKG_BUILD ON)
set (VCPKG_TARGET_TRIPLET "x64-windows-static" CACHE STRING "")
set (VCPKG_FEATURE_FLAGS "manifests,binarycaching,versions")
set (ENV{VCPKG_BINARY_SOURCES} "nuget,github,readwrite")
set (ENV{VCPKG_USE_NUGET_CACHE} "true")
set (ENV{VCPKG_MANIFEST_INSTALL} "ON")
set (ENV{VCPKG_BUILD} "ON")
set (ENV{VCPKG_FEATURE_FLAGS} "manifests,binarycaching,versions")
set (ENV{VCPKG_TARGET_TRIPLET} "x64-windows-static")
```
One piece of output seems odd is the inclusion of a -DVCPKG_MANIFEST_INSTALL=OFF parameter. As shown above, the CMakeLists.txt file is setting that variable and environment variable ON, yet the output shows it being OFF.
```ps1
[DEBUG] 1001: CreateProcessW("C:\Program Files\CMake\bin\cmake.exe" -DVCPKG_ROOT_DIR=C:/Users/frederick.brier/Projects/WzRefactor/WzLib-org/vcpkg -DPACKAGES_DIR=C:/Users/frederick.brier/Projects/WzRefactor/WzLib-org/vcpkg/packages -DBUILDTREES_DIR=C:/Users/frederick.brier/Projects/WzRefactor/WzLib-org/vcpkg/buildtrees -D_VCPKG_INSTALLED_DIR=C:/Users/frederick.brier/Projects/WzRefactor/WzLib-org/vcpkg_installed -DDOWNLOADS=C:/Users/frederick.brier/Projects/WzRefactor/WzLib-org/vcpkg/downloads -DVCPKG_MANIFEST_INSTALL=OFF -P "C:\Users\frederick.brier\Projects\WzRefactor\WzLib-org\vcpkg\buildtrees\0.vcpkg_dep_info.cmake")
```

**Reply by Osyotr** [reply, marked as answer]
It is not clear what your question is but I assume you just want to setup a project that uses vcpkg in manifest mode.

> `VCPKG_INSTALLED_DIR=C:/Users/frederick.brier/Projects/WzRefactor/WzLib-org/vcpkg_installed`

This is the folder where packages will be installed in manifest mode. You don't need to set additional variables except `CMAKE_TOOLCHAIN_FILE`, `VCPKG_TARGET_TRIPLET` and maybe `VCPKG_HOST_TRIPLET` and that should happen before first `project(...)` call. Don't forget to clear CMake cache if `CMAKE_TOOLCHAIN_FILE` has been changed.

**Reply by dg0yt** [reply]
With CMake, you should simply not use command-line `vcpkg install` which operates in the source dir's `vcpkg_installed`. Leave everything to the vcpkg toolchain file and the triplet variables: During CMake configuration, it will put packages in the project's binary dir, and this is a good thing at least if you deal with multiple targets from the same source dir.

**Reply by frederick-brier** [reply]
@Osyotr Thank you so much. I never would have thought to place the variables above the project(...). @dg0yt Thank you for being clear about removing the calls to the command line. If you have a place to put articles, I would be glad to write one on using cmake/vcpkg/vcpkg.json/boost with Github Packages as a binary cache in a private repo.

I did have one last issue with boost::filesystem crashing the build, but decided trying to override the version to 1.67.0 (as per the vcpkg.json above) was just one reach too far. I removed that JSON and the build succeeded. Again, thank you SO much.

In summary, I placed the following lines before the project(...) , removed the execute_process(...) calls and commented out the other VCPKG_* variables:
```cmake
set (CMAKE_TOOLCHAIN_FILE "${CMAKE_SOURCE_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake")
set (VCPKG_INSTALLED_DIR "${CMAKE_SOURCE_DIR}/vcpkg_installed")
set (VCPKG_TARGET_TRIPLET "x64-windows-static" CACHE STRING "")
set (VCPKG_HOST_TRIPLET "x64-windows-static" CACHE STRING "")
set (VCPKG_BINARY_SOURCES "nuget,github,readwrite")
set (VCPKG_USE_NUGET_CACHE "true")
```
I also changed these variables to reflect what I have come to understand is the Boost_ROOT:
```cmake
set (BOOST_ROOT "${CMAKE_SOURCE_DIR}/vcpkg_installed/x64-windows-static")
set (BOOST_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/vcpkg_installed/x64-windows-static/include")
set (BOOST_LIBRARY_DIR "${CMAKE_SOURCE_DIR}/vcpkg_installed/x64-windows-static/lib")
```

**Reply by Osyotr** [reply]
No need to set boost variables, just use CMake's `FindBoost.cmake`:
`find_package(Boost REQUIRED COMPONENTS filesystem ...)` and `target_link_libraries(... PRIVATE Boost::headers Boost::filesystem ...)`

**Reply by dg0yt** [reply]
This shouldn't be used when dealing with different target triplets / cmake binary dirs for the same source dir:
`set (VCPKG_INSTALLED_DIR "${CMAKE_SOURCE_DIR}/vcpkg_installed")`

**Comment by frederick-brier** [comment]
I removed the Boost variables. Thank you.
